### ISSUE TITLE
Duplicated identifier used when mangle is enabled

### ISSUE THREAD

**Issue by renanccastro** [terser/terser#1155]
**Bug report**

Terser is using the same identifier for the function, and for the const defined inside the if. This produces code that can't be evaluated.

Here is a simple reproduction case:

```js
function f1() {
	if (true) {
		function f2() {
			return 1;
		}

		const var1 = 666;
	}
}
```

If var1 is outside the "if", the correct identifier is used when minifying.

Used options:

```js
{
  module: true,
  compress: {
    unused: false
  }
}
```

**Curent terser output**

```js
function n(){{function n(){return 1}const n=666}}
```

note the inner function n() and the const n, they clash.

**Comment by fabiosantoscode** [comment]
Terser really can't deal with scopes of function declarations when they're inside of a block. [This section](https://tc39.es/ecma262/multipage/additional-ecmascript-features-for-web-browsers.html#sec-block-level-function-declarations-web-legacy-compatibility-semantics) of the spec isn't being properly followed.

I'm thinking that this pattern should trigger the scope's `pinned` property, so that inlining and dead code elimination is no longer attempted in these scopes.

I hate to learn that this ancient issue caused grief to someone.

**Comment by fabiosantoscode** [comment]
Note that declaring `'use strict'`, fixes the problem too. So this is really 2 bugs in one as `module: true` isn't being followed.

**Comment by fabiosantoscode** [comment]
I've been looking around and trying different things. I can fix this but haven't decided what's the best long term fix. To fix the case of section B 3.2.1 (funcdefs in block scopes) or avoid mangling 2 things in the same function to the same name altogether, which fixes all the cases in that section?

Since block scope conflicts are often causes of bugs (and incompatibility with older Safari) I'm thinking that the mangler should avoid reusing any names in the same scope. This will, though, make the output slightly larger.
